src/Quadrature.cpp

```diff
@@ -1184,6 +1110,171 @@ void GLPolarQuad::precomputeWeights(bool solve_3D) {
 
 
+/** 
+ * @brief    the Legendre polynomial of degree n evaluated at x
+ * @param    n an integer >=0: the order of the polynomial
+ * @param    x in (-1,1), the point at which to evaluate the polynomial
+ * @return   the value of the Legendre polynomial of degree n at x
+ */
+double GLPolarQuad::legendrePolynomial(int n, double x) {
+  if (n == 0)
+    return 1;
+  if (n == 1)
+    return x;
+  else {
+    double c = 2.0*(n-2) + 3.0;
+    double b = 1.0*(n-2) + 2.0;
+    double a = 1.0*(n-2) + 1.0;
+    double value =
+      c/b * x * legendrePolynomial(n-1, x) - a/b * legendrePolynomial(n-2,x);
+    return value;
+  }
+}
+ 
+
+/**
+ * @brief    the first logarithmic derivative of a Legendre polynomial
+ * @param    m the order of the polynomial
+ * @param    x point at which to evaluate the logarithmic derivative
+ * @return   the value of the logarithmic derivative at x
+ */
+double GLPolarQuad::logDerivLegendre(int n, double x) {
+  double num = n * x - n * legendrePolynomial(n-1,x) / legendrePolynomial(n,x);
+  double denom = x*x - 1;
+  return num/denom;
+}
+
+
+/**
+ * @brief    the second logarithmic derivative of a Legendre polynomial
+ * @param    m the order of the polynomial
+ * @param    x point at which to evaluate the logarithmic derivative
+ * @return   the value of the logarithmic derivative at x
+ */
+double GLPolarQuad::secondLogDerivLegendre(int n, double x) {
+  double num =
+    n*(n+1) + logDerivLegendre(n,x) * ((1-x*x)* logDerivLegendre(n,x) - 2 * x);
+  double denom = x*x-1;
+  return num/denom;
+}
+  
+
+/**
+ * @brief    finds the roots of Legendre polynomial of order n
+ * @detail   guesses for positive roots are set at logarithmic intervals. 
+ *           Positive roots are found simultaneously using an 
+ *           Alberth-Householder-n method. Each guess is successively nudged
+ *           towards a true root. Only the positive roots are calculated
+ * @param    n the order of the polynomial
+ * @return   a list of the roots of the polynomial
+ */
+std::vector <double> GLPolarQuad::getLegendreRoots(int n) {
+
+  /* put these somewhere else */
+  double E1 = 1e-10;
+  double E2 = 1e-10;
+
+  std::vector <double> roots;
+  std::vector <bool> converged;
+  std::vector <double> s1_tilde;
+  std::vector <double> s2_tilde;
+
+  /* set guesses with log scale*/
+  for (int i=0; i < n/2; ++i) {
+    roots.push_back(- pow(2, (-.5*(i+1))) +1);
+    converged.push_back(false);
+    s1_tilde.push_back(0);
+    s2_tilde.push_back(0);
+
+  }
+
+  if (n%2 == 1) {
+    roots.push_back(0);
+    converged.push_back(true);
+    s1_tilde.push_back(0);
+    s2_tilde.push_back(0);
+  }
+
+
+  bool all_roots_converged = false;
+
+  /* use the Alberth-Housholder_n method to nudge guesses towards roots */
+  while (not all_roots_converged) {
+    
+    /* set S tildes */
+    for (int i=0; i < (n+1)/2; ++i) {
+      if (not converged[i]) {
+        double sum1 = 0;
+        double sum2 = 0;
+        for (int j=0; j<= (n+1)/2; ++j) {
```

Comment:
There is an inconsistency here. Your vectors were only made to of size `(n+1)/2`. Here you go to element `(n+1)/2` (ie assuming that the vector has size `(n+1)/2+1`. This is the reason why your tests are failing. It's picking up a random value in that last position. By the nature of how `std::vector` works, that value is often zero so you weren't seeing the issue. But that value is not guaranteed to be zero. And in the test being run, that memory is being reused so it essentially picks up a random value. If you access with `.at(...)` instead of `[]` the bounds error will show up right away.